src/core/server/elasticsearch/client/configure_client.ts

```diff
@@ -137,14 +139,46 @@ const addLogging = (client: Client, logger: Logger) => {
             http: { request: { id: event.meta.request.options.opaqueId } },
           }
         : undefined; // do not clutter logs if opaqueId is not present
+      let queryMessage = '';
       if (error) {
         if (error instanceof errors.ResponseError) {
-          logger.debug(`${getResponseMessage(event)} ${getErrorMessage(error)}`, meta);
+          queryMessage = `${getResponseMessage(event)} ${getErrorMessage(error)}`;
         } else {
-          logger.debug(getErrorMessage(error), meta);
+          queryMessage = getErrorMessage(error);
         }
       } else {
-        logger.debug(getResponseMessage(event), meta);
+        queryMessage = getResponseMessage(event);
+      }
+
+      queryLogger.debug(queryMessage, meta);
+
+      if (event.warnings && event.warnings.length > 0) {
+        // Plugins can explicitly mark requests as originating from a user by
+        // removing the `'x-elastic-product-origin': 'kibana'` header that's
+        // added by default. User requests will be shown to users in the
+        // upgrade assistant UI as an action item that has to be addressed
+        // before they upgrade.
+        // Kibana requests will be hidden from the upgrade assistant UI and are
+        // only logged to help developers maintain their plugins
+        const requestOrigin =
+          (event.meta.request.options.headers != null &&
+            (event.meta.request.options.headers[
+              'x-elastic-product-origin'
+            ] as unknown as string)) === 'kibana'
+            ? 'kibana'
+            : 'user';
+
+        const stackTrace = new Error().stack?.split('\n').slice(5).join('\n');
+
+        if (requestOrigin === 'kibana') {
+          deprecationLogger.warn(
+            `ES DEPRECATION: ${event.warnings}\nOrigin:${requestOrigin}\nStack trace:\n${stackTrace}\nQuery:\n${queryMessage}`
```

Comment:
Do we want to include all of this info in the log message like we do for query logs? Or would it make sense to move some of it into `LogMeta`?

I suppose these are mostly intended for devs, who will likely be reading it out of a console, but I guess I'm thinking about how messy the messages will look in json layout.

Reply:
Also nit - maybe we could pull the message out into a const to make it clearer that the only difference here is whether we are logging a warning or an error?

Reply:
yeah, we could use logmeta for 8.x but on 7.x most users won't see the logMeta (especially cloud). If this is disabled by default it probably doesn't matter too much either way... but like you say for a dev and for CI the text output with newlines is more readable.

And if we decide to enable it by default so that we could surface this on Cloud then we loose the usefulness if we loose the logmeta like stack and complete query.

I am starting to doubt enabling by default though as it could potentially create a lot of log volume. We could consider making user deprecations log on `debug` but kibana origin deprecations log on `warn` so that by default the log volume should be 0 unless we have a bug. That would give us high confidence from 7.16 users on cloud that we've fixed deprecations.

And when users upgrade they can anyway see the user deprecations in the upgrade assistant UI, so that's probably a second reason not to log this by default.

Reply:
> We could consider making user deprecations log on debug but kibana origin deprecations log on warn so that by default the log volume should be 0 unless we have a bug. That would give us high confidence from 7.16 users on cloud that we've fixed deprecations.

That seems like a reasonable alternative to me.